### DISCUSSION TITLE
Lesson4: Rinkeby store transaction failed

### DISCUSSION THREAD

**Discussion by CJ-Rose** [smartcontractkit/full-blockchain-solidity-course-py#23] [Q&A] [answered]
Using Infura to deploy to Rinkeby: 
Inspection of transactions on Etherscan shows contract creation successful, my transaction details match the video. The second transaction failed bc it ran out of gas. Looks like it had about half of what it needed. I haven't been able to find where the gas limit is being set. But I deployed again and it went through, using about twice as much gas as was available the first time, and the same as in the video. Any idea what happened?

**Comment by PatrickAlphaC** [comment, marked as answer]
EDIT: Fixed: Just rerun the script! Likely a testnet issue. 

Could you add the code and error text of the issue?

**Reply by CJ-Rose** [reply]
<img width="522" alt="Screen Shot 2021-09-16 at 9 53 15 AM" src="https://user-images.githubusercontent.com/79813968/133653576-fffbb00d-0d8e-46f8-bf69-474c4c378ce6.png">

<img width="635" alt="Screen Shot 2021-09-16 at 9 56 08 AM" src="https://user-images.githubusercontent.com/79813968/133653826-25bc15b5-4824-4e60-89d9-791679b089a9.png">


Didn't get an error in my terminal. You can see the first time I deployed: printed "Updated!" but returned 0 from call to retrieve().
Second try returned 15 and Etherscan showed success.

<img width="612" alt="Screen Shot 2021-09-16 at 9 55 03 AM" src="https://user-images.githubusercontent.com/79813968/133653845-0cfe067b-2449-402f-b7a0-ae63ac0c8fc1.png">

**Reply by AlbertoSinigaglia** [reply]
is this consistent or it happened only that time? (also, for the following times that you will post code, please post the source code and not images, so we can try it without asking to you)

**Reply by PatrickAlphaC** [reply]
Could you add it in text rather than images? Its a lot harder to copy paste your code as-is. And send a link to that Etherscan transaction?

**Reply by PatrickAlphaC** [reply]
What's in your `SimpleStorage.sol` file as well please!

**Reply by CJ-Rose** [reply]
This only happened the first time. I have since been able to interact with the contract and continue on with the course. Just trying to understand why this happened. And I've just now learned how to add screenshots, code, and links to Github, so thanks for helping me to do it right and for being patient : )


my deploy.py:
```python
from solcx import compile_standard, install_solc
import json
from web3 import Web3
import os
from dotenv import load_dotenv

load_dotenv()

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

print("Installing...")
install_solc("0.6.0")


# Solidity source code
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                }
            }
        },
    },
    solc_version="0.6.0",
)

# write compiled code to new file
with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

# get bytecode
bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

# get ABI
abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

# connect to Rinkeby
w3 = Web3(
    Web3.HTTPProvider("https://rinkeby.infura.io/v3/74bf2da4c26c48559ad9f61aee7966b5")
)
chain_id = 4
my_address = "0xae9AC410a5597f2Ca3DDF1757FCD516DaE9AC298"
private_key = os.getenv("PRIVATE_KEY")


# create the contract in python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)
# get the lastest transaction
nonce = w3.eth.getTransactionCount(my_address)
# build transaction
transaction = SimpleStorage.constructor().buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce}
)
# sign transaction
signed_txn = w3.eth.account.sign_transaction(transaction, private_key)
# send transaction
print("Deploying contract...")
txn_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
txn_receipt = w3.eth.wait_for_transaction_receipt(txn_hash)
print("Deployed!")

# working with the contract (need contract address, contract ABI)
simple_storage = w3.eth.contract(address=txn_receipt.contractAddress, abi=abi)
# call -> simulate a call and gets a return value
#  transact -> make a state change
#  print initial value of favorite number (0)
print(simple_storage.functions.retrieve().call())

#  build, sign, send transaction
print("Updating Contract...")
store_transaction = simple_storage.functions.store(20).buildTransaction(
    {"chainId": chain_id, "from": my_address, "nonce": nonce + 1}
)
signed_store_txn = w3.eth.account.sign_transaction(store_transaction, private_key)
send_store_txn = w3.eth.send_raw_transaction(signed_store_txn.rawTransaction)
store_txn_receipt = w3.eth.wait_for_transaction_receipt(send_store_txn)
print("Updated!")
print(simple_storage.functions.retrieve().call())
```


my SimpleStorage.sol:
```javascript
// SPDX-License-Identifier: MIT

pragma solidity ^0.6.0;

contract SimpleStorage {
    //initialize to 0
    uint256 favoriteNumber;

    struct People {
        uint256 favoriteNumber;
        string name;
    }

    People[] public people;
    mapping(string => uint256) public nameToFavoriteNumber;

    function store(uint256 _favoriteNumber) public returns (uint256) {
        favoriteNumber = _favoriteNumber;
        return _favoriteNumber;
    }

    function retrieve() public view returns (uint256) {
        return favoriteNumber;
    }

    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }
}
```

[failed tx on Etherscan](https://rinkeby.etherscan.io/tx/0x3e9c9afb5309e1851d4c92f1937a2ea02a02c6d54f53edfcf27745ee990ca0fe)

**Reply by PatrickAlphaC** [reply]
Thank you for all this information!! I've tested, and it looks like it works fine. It could just have been an issue with the rinkeby testnet (testnets aren't as good as mainnet). I'll monitor to see if anyone else runs into this issue. 

Also, could you please get a new API key from Infura? You've exposed yours now so someone could make you hit the upper limit on calls (meaning, you'd not be able to send future transactions)! 

Aka, get a new url from this line:
```
w3 = Web3(
    Web3.HTTPProvider("XXXX")
)
```

Don't worry, you didn't send your private keys or anything in that message, so all testnet funds are safe.

**Reply by CJ-Rose** [reply]
got it, thank you!
